File: legacy/python/src/roostery/agent_detect.py
"""agent_detect — 检测本机装了哪些 AI agent CLI（CC / Codex / Gemini）。

产品理念：roostery init 不强制要求用户列清单，自动检测后只 hook 装了的。
没装的不强求；装了但用户不想 hook 的，提供 ``--skip-agent`` 覆盖。

依据 memory ``agent_hooks_facts.md``（必读，更新随 agent CLI 升级）：
- **CC**：``~/.claude/settings.json``，事件 ``SessionEnd``（per-session，非 Stop）
- **Codex**：本身**无**用户 hook；``~/.codex/hooks.json`` 实际是
  ``codex-plugin-cc`` 这个 CC 插件用的，跟 CC 一样的 hook 格式。我们仍 merge
  到这个文件——用户装了 codex-plugin-cc 才能用上，没装也不会破坏。
- **Gemini CLI**：``~/.gemini/settings.json``，事件 ``SessionEnd``；首选
  ``AfterAgent`` 因为 stdin 直接含 ``prompt_response``，但 per-turn 会刷屏，
  所以这里也用 ``SessionEnd``。
"""
from __future__ import annotations

import os
import shutil
from dataclasses import dataclass
from pathlib import Path
from typing import List, Optional


@dataclass(frozen=True)
class AgentSpec:
    """一种 AI agent CLI 的检测 + hook 配置参数。"""
    name: str                 # 显示名："cc" / "codex" / "gemini"
    cli: str                  # 探测用的可执行命令名："claude" / "codex" / "gemini"
    template: str             # roostery/templates/ 下的 .json.tmpl 文件名
    hooks_target: Path        # 用户主目录下的目标 hook 配置文件路径


def _expand(p: str) -> Path:
    return Path(os.path.expanduser(p))


AGENTS: List[AgentSpec] = [
    AgentSpec(
        name="cc",
        cli="claude",
        template="claude_code_settings.json.tmpl",
        hooks_target=_expand("~/.claude/settings.json"),
    ),
    AgentSpec(
        name="codex",
        cli="codex",
        template="codex_hooks.json.tmpl",
        hooks_target=_expand("~/.codex/hooks.json"),
    ),
    AgentSpec(
        name="gemini",
        cli="gemini",
        template="gemini_settings.json.tmpl",
        hooks_target=_expand("~/.gemini/settings.json"),
    ),
]


@dataclass(frozen=True)
class DetectResult:
    """检测结果。``cli_path`` None 表示未装。"""
    spec: AgentSpec
    cli_path: Optional[str]

    @property
    def installed(self) -> bool:
        return self.cli_path is not None


def detect_all(*, skip: Optional[List[str]] = None) -> List[DetectResult]:
    """检测全部 ``AGENTS``。``skip=["codex"]`` 跳过 codex 强制 unknown。"""
    skip_set = set(skip or [])
    out: List[DetectResult] = []
    for spec in AGENTS:
        if spec.name in skip_set:
            out.append(DetectResult(spec=spec, cli_path=None))
            continue
        path = shutil.which(spec.cli)
        out.append(DetectResult(spec=spec, cli_path=path))
    return out


def installed_only(results: List[DetectResult]) -> List[DetectResult]:
    """过滤出真装了的。"""
    return [r for r in results if r.installed]


def describe(results: List[DetectResult]) -> str:
    """格式化检测结果给人看。"""
    lines = []
    for r in results:
        mark = "✓" if r.installed else "—"
        path = r.cli_path or "(not installed)"
        lines.append(f"  [{mark}] {r.spec.name:7} | cli={path}")
    return "\n".join(lines)

File: crates/roostery/src/agent_detect.rs
//! Detect installed AI agent CLIs (CC / Codex / Gemini) on the host.
//!
//! Used by `roostery init` to decide which Stop hook templates to merge.
//! `which::which()` does PATH-walk lookup; not-found is **not** an error
//! (returns `cli_path: None`).
//!
//! See `.codestable/features/2026-05-18-roostery-init/roostery-init-design.md`
//! §2.1.2.

use crate::hooks_merge::AgentKind;
use std::path::PathBuf;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[non_exhaustive]
pub struct AgentSpec {
    pub kind: AgentKind,
    /// Executable name on PATH (e.g. `"claude"`).
    pub cli: &'static str,
    /// Hooks file target, with `~` un-expanded (consumer expands via
    /// [`AgentSpec::expanded_hooks_target`]).
    pub hooks_target: &'static str,
}

pub const AGENTS: &[AgentSpec] = &[
    AgentSpec {
        kind: AgentKind::Cc,
        cli: "claude",
        hooks_target: "~/.claude/settings.json",
    },
    AgentSpec {
        kind: AgentKind::Codex,
        cli: "codex",
        hooks_target: "~/.codex/hooks.json",
    },
    AgentSpec {
        kind: AgentKind::Gemini,
        cli: "gemini",
        hooks_target: "~/.gemini/settings.json",
    },
];

impl AgentSpec {
    /// Resolve `~/<rest>` to `<home>/<rest>`. Falls back to literal string
    /// if HOME is unavailable.
    pub fn expanded_hooks_target(self) -> PathBuf {
        if let Some(rest) = self.hooks_target.strip_prefix("~/")
            && let Some(home) = dirs::home_dir()
        {
            return home.join(rest);
        }
        PathBuf::from(self.hooks_target)
    }
}

#[derive(Debug, Clone)]
pub struct DetectResult {
    pub spec: AgentSpec,
    /// `Some(path)` if CLI was found on PATH, `None` otherwise (incl. skipped).
    pub cli_path: Option<PathBuf>,
}

impl DetectResult {
    pub fn installed(&self) -> bool {
        self.cli_path.is_some()
    }
}

/// Detect all known agents. `skip` forces those agents to be reported as
/// not installed even if their CLI is on PATH.
pub fn detect_all(skip: &[AgentKind]) -> Vec<DetectResult> {
    AGENTS
        .iter()
        .copied()
        .map(|spec| {
            let cli_path = if skip.contains(&spec.kind) {
                None
            } else {
                which::which(spec.cli).ok()
            };
            DetectResult { spec, cli_path }
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn agents_const_has_three_entries() {
        assert_eq!(AGENTS.len(), 3);
        assert_eq!(AGENTS[0].kind, AgentKind::Cc);
        assert_eq!(AGENTS[1].kind, AgentKind::Codex);
        assert_eq!(AGENTS[2].kind, AgentKind::Gemini);
    }

    #[test]
    fn expanded_hooks_target_resolves_tilde() {
        let cc = AGENTS[0];
        let p = cc.expanded_hooks_target();
        assert!(!p.to_string_lossy().starts_with('~'));
        assert!(p.ends_with(".claude/settings.json"));
    }

    #[test]
    fn skip_forces_not_installed_regardless_of_path() {
        let results = detect_all(&[AgentKind::Cc, AgentKind::Codex, AgentKind::Gemini]);
        assert_eq!(results.len(), 3);
        for r in &results {
            assert!(!r.installed(), "skipped agent must report not-installed");
        }
    }

    #[test]
    fn detect_all_returns_three_results_with_correct_order() {
        let results = detect_all(&[]);
        assert_eq!(results.len(), 3);
        assert_eq!(results[0].spec.kind, AgentKind::Cc);
        assert_eq!(results[1].spec.kind, AgentKind::Codex);
        assert_eq!(results[2].spec.kind, AgentKind::Gemini);
    }
}
